### ISSUE TITLE
`using namespace` overrides previous `using namespace` uses as interactive

### ISSUE THREAD

**Issue by helpimnotdrowning** [PowerShell/PowerShell#23873]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

Run the following code in an interactive session:
```pwsh
using namespace System.Net.Http.Headers
using namespace System.Collections.Generic

[queue[string]] ; [ProductInfoHeaderValue]
```

### Expected behavior

```console
Both types should be usable, the following should output:

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Queue`1                                  System.Object
True     False    ProductInfoHeaderValue                   System.Object
```

A workaround is usable, where all `using namespace` statements are written on the same line, separated by semicolons, like `using namespace System.Net.Http.Headers; using namespace System.Collections.Generic`, which shows the expected output.
```


### Actual behavior

```console
Depending on the order the `using namespace` statements are used, the following outputs appear:

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Queue`1                                  System.Object
InvalidOperation: Unable to find type [ProductInfoHeaderValue].
```
or 
```
InvalidOperation: Unable to find type [queue].

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    ProductInfoHeaderValue                   System.Object
```
```


### Error details

```console
Exception             :
    Type        : System.Management.Automation.RuntimeException
    ErrorRecord :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Unable to find type [ProductInfoHeaderValue].
            HResult : -2146233087
        TargetObject          : ProductInfoHeaderValue
        CategoryInfo          : InvalidOperation: (ProductInfoHeaderValue:TypeName) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : TypeNotFound
        InvocationInfo        :
            ScriptLineNumber : 1
            OffsetInLine     : 19
            HistoryId        : 39
            Line             : [queue[string]] ; [ProductInfoHeaderValue]
            Statement        : [ProductInfoHeaderValue]
            PositionMessage  : At line:1 char:19
                               + [queue[string]] ; [ProductInfoHeaderValue]
                               +                   ~~~~~~~~~~~~~~~~~~~~~~~~
            CommandOrigin    : Internal
        ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
    TargetSite  :
        Name          : ResolveTypeName
        DeclaringType : [System.Management.Automation.TypeOps]
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message     : Unable to find type [ProductInfoHeaderValue].
    Data        : System.Collections.ListDictionaryInternal
    Source      : System.Management.Automation
    HResult     : -2146233087
    StackTrace  :
   at System.Management.Automation.TypeOps.ResolveTypeName(ITypeName typeName, IScriptExtent errorPos)
   at System.Management.Automation.Interpreter.FuncCallInstruction`3.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
TargetObject          : ProductInfoHeaderValue
CategoryInfo          : InvalidOperation: (ProductInfoHeaderValue:TypeName) [], RuntimeException
FullyQualifiedErrorId : TypeNotFound
InvocationInfo        :
    ScriptLineNumber : 1
    OffsetInLine     : 19
    HistoryId        : 39
    Line             : [queue[string]] ; [ProductInfoHeaderValue]
    Statement        : [ProductInfoHeaderValue]
    PositionMessage  : At line:1 char:19
                       + [queue[string]] ; [ProductInfoHeaderValue]
                       +                   ~~~~~~~~~~~~~~~~~~~~~~~~
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.2
PSEdition                      Core
GitCommitId                    7.4.2
OS                             Microsoft Windows 10.0.26120
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![image](https://github.com/PowerShell/PowerShell/assets/35247379/5dd28e85-3067-420e-b788-94e602625acc)

**Comment by helpimnotdrowning** [comment]
Related: #19207 , was closed due to inactivity

**Comment by MartinGC94** [comment]
It works fine if you paste it in as one big block:
![image](https://github.com/PowerShell/PowerShell/assets/42123497/90f92d0e-2785-46ef-becd-eaff8862c896)
If you paste with Ctrl+v the pasting is handled by PSReadLine but if you right click to paste it's the console that handles the pasting and it just types out each character one by one so when you get a linebreak it submits the script text as is.

Note this is with the old consolehost, I think in the new Windows terminal you need to unbind Ctrl+V in the terminal settings so the key presses reaches PowerShell/PSReadline.

**Comment by mklement0** [comment]
Pasting as a single, multiline piece of code is definitely a _workaround_ (and in Windows Terminal an especially cumbersome one), but I think the larger issue is the questionable behavior of the most recently submitted `using namespace` statement _undoing all previous ones_.

E.g., if you later submit `using namespace System.Windows.Forms`, all previous `using namespace` statements are forgotten.

**Comment by dwtaber** [comment]
> I expect most instances of `using namespace` in the console to be used when pasting chunks of code that happen to include them. For that scenario it seems risky to keep old `using namespace` references because you might run into type name conflicts due to some other snippet the user may have run earlier.

I have some `using namespace` statements in my profile script for convenience, and it's a bit annoying that they vanish if I enter another one from the console. It's worth noting that dot-sourcing a script with `using namespace` statements also overrides prior statements.

Is there any way to enumerate the `using namespace` statements currently in effect? That seems like it would be handy, especially if it were also possible to add _or remove_ specific namespaces rather than override them all every time.

**Comment by MartinGC94** [comment]
> I have some using namespace statements in my profile script for convenience, and it's a bit annoying that they vanish if I enter another one from the console.

I understand that it's annoying to lose the old using statements, but if it's the scenario I'm describing then IMO it's even more annoying if a pasted piece of code doesn't work as expected.

> Is there any way to enumerate the using namespace statements currently in effect?

I have an old saved snippet for this:
```
$flags                = [System.Reflection.BindingFlags]'Instance, NonPublic'
$internalSessionState = $ExecutionContext.SessionState.GetType().GetProperty('Internal', $flags).GetValue($ExecutionContext.SessionState)
$globalScope          = $internalSessionState.GetType().GetProperty('GlobalScope', $flags).GetValue($internalSessionState)
$typeResolutionState  = $globalScope.GetType().GetProperty('TypeResolutionState', $flags).GetValue($globalScope)
$typeResolutionState.GetType().GetField('namespaces', $flags).GetValue($typeResolutionState)
```

**Comment by mklement0** [comment]
Generally, what you do interactively in the console affects the remainder of the session (if done in the global scope).

E.g., setting a variable, environment variable (scope-independently), changing preference variables, ...

That `.ps1` script calls therefore lack isolation from the interactive environment - unlike in POSIX-compatible shells, where scripts always run in _child_ processes, is a fundamental pitfall - but that's a separate discussion.

To me it makes for an awkward inconsistency not to treat `using namespace` calls as staying in effect for the remainder of the session, and @dwtaber's use case is an example of that awkwardness.

In other words: The consistency and convenience argument to me outweighs the concern over lingering `using namespace` statements affecting later code.

**Comment by dwtaber** [comment]
> I understand that it's annoying to lose the old using statements, but if it's the scenario I'm describing then IMO it's even more annoying if a pasted piece of code doesn't work as expected.
> 
On the other hand, someone unaware of the current behavior would probably be confused about their pasted-in code not working as expected if they wrote it in the belief that the prior `using namespace` statements would still be in effect. Since either discarding prior statements or keeping them has the potential to cause unexpected behavior, my preference would be for prior statements to persist, and to try to identify and warn about any name collisions introduced by subsequent statements.

> > Is there any way to enumerate the using namespace statements currently in effect?
> 
> I have an old saved snippet for this:
> 
Very cool, thank you!
